File: PathFinder/backend/algorithms/astar.py
"""
A* (A-Star) algorithm implementation for the pathfinding visualizer.

Functions:
    astar(grid, start, end):
        - grid: 2D list of ints where 0 = empty cell, 1 = wall
        - start: tuple (row, col) for the starting cell
        - end: tuple (row, col) for the ending cell
        Returns a tuple (visited_order, path):
            visited_order: list of [row, col] in the order nodes are dequeued (first visit)
            path: list of [row, col] forming the shortest path from start to end (inclusive),
                  or empty list if no path exists

Uses Manhattan distance as the heuristic.
"""
import heapq


def astar(grid, start, end):
    rows = len(grid)
    cols = len(grid[0]) if rows > 0 else 0

    def in_bounds(r, c):
        return 0 <= r < rows and 0 <= c < cols

    def heuristic(a, b):
        # Manhattan distance
        return abs(a[0] - b[0]) + abs(a[1] - b[1])

    # Validate start/end
    if not in_bounds(*start) or not in_bounds(*end):
        return [], []
    if grid[start[0]][start[1]] == 1 or grid[end[0]][end[1]] == 1:
        return [], []

    # g_score: cost from start to node
    g_score = { (r, c): float('inf') for r in range(rows) for c in range(cols) }
    g_score[start] = 0

    # f_score: g_score + heuristic
    f_score = { (r, c): float('inf') for r in range(rows) for c in range(cols) }
    f_score[start] = heuristic(start, end)

    # Priority queue: (f_score, count, node)
    open_heap = []
    heapq.heappush(open_heap, (f_score[start], 0, start))
    open_set = {start}

    parent = {}
    visited = set()
    visited_order = []
    count = 0

    # Neighbor directions: up, right, down, left
    directions = [(-1, 0), (0, 1), (1, 0), (0, -1)]

    while open_heap:
        _, _, current = heapq.heappop(open_heap)
        open_set.discard(current)

        # Skip if already visited
        if current in visited:
            continue

        visited.add(current)
        visited_order.append([current[0], current[1]])

        # Check if reached end
        if current == end:
            break

        # Explore neighbors
        for dr, dc in directions:
            nr, nc = current[0] + dr, current[1] + dc
            neighbor = (nr, nc)
            if in_bounds(nr, nc) and neighbor not in visited and grid[nr][nc] == 0:
                tentative_g = g_score[current] + 1
                if tentative_g < g_score[neighbor]:
                    parent[neighbor] = current
                    g_score[neighbor] = tentative_g
                    f_score[neighbor] = tentative_g + heuristic(neighbor, end)
                    if neighbor not in open_set:
                        count += 1
                        heapq.heappush(open_heap, (f_score[neighbor], count, neighbor))
                        open_set.add(neighbor)

    # Reconstruct path
    path = []
    if end in parent or start == end:
        node = end
        while node != start:
            path.append([node[0], node[1]])
            node = parent.get(node)
            if node is None:
                break
        path.append([start[0], start[1]])
        path.reverse()

    return visited_order, path

File: AutoDriver/src/core/astar.cpp
#include "core/astar.h"
#include <queue>
#include <unordered_map>
#include <algorithm>
#include <cmath>

AStar::AStar(const Grid& grid) : grid_(grid) {}

AStarResult AStar::findPath(Vec2i start, Vec2i goal) {
    AStarResult result;
    
    // Validate start and goal
    if (!grid_.isValid(start.x, start.y) || grid_.isObstacle(start.x, start.y) ||
        !grid_.isValid(goal.x, goal.y) || grid_.isObstacle(goal.x, goal.y)) {
        return result;  // Invalid positions
    }
    
    // Priority queue for open set (min-heap by f_cost)
    auto cmp = [](Node* a, Node* b) { return *a > *b; };
    std::priority_queue<Node*, std::vector<Node*>, decltype(cmp)> open_set(cmp);
    
    // Track visited nodes
    std::unordered_set<Vec2i, Vec2iHash> closed_set;
    
    // Map from position to node (for finding existing nodes)
    std::unordered_map<Vec2i, Node*, Vec2iHash> node_map;
    
    // Storage for all allocated nodes (for cleanup)
    std::vector<std::unique_ptr<Node>> all_nodes;
    
    // Create start node
    auto start_node = std::make_unique<Node>(
        start, 
        0.0f, 
        euclideanDistance(start, goal),
        nullptr
    );
    Node* start_ptr = start_node.get();
    node_map[start] = start_ptr;
    open_set.push(start_ptr);
    all_nodes.push_back(std::move(start_node));
    
    Node* goal_node = nullptr;
    
    while (!open_set.empty()) {
        Node* current = open_set.top();
        open_set.pop();
        
        // Skip if already visited
        if (closed_set.count(current->pos)) {
            continue;
        }
        
        // Mark as visited
        closed_set.insert(current->pos);
        result.visited.push_back(current->pos);
        result.nodes_expanded++;
        
        // Check if we reached the goal
        if (current->pos == goal) {
            goal_node = current;
            result.success = true;
            result.path_cost = current->g_cost;
            break;
        }
        
        // Explore neighbors
        for (const Vec2i& neighbor_pos : getNeighbors(current->pos)) {
            if (closed_set.count(neighbor_pos)) {
                continue;
            }
            
            // Calculate costs
            float tentative_g = current->g_cost + 1.0f;  // Uniform cost
            
            // Check if we already have this neighbor in open set
            auto it = node_map.find(neighbor_pos);
            if (it != node_map.end()) {
                // Update if we found a better path
                if (tentative_g < it->second->g_cost) {
                    it->second->g_cost = tentative_g;
                    it->second->parent = current;
                }
            } else {
                // Create new node
                auto neighbor_node = std::make_unique<Node>(
                    neighbor_pos,
                    tentative_g,
                    euclideanDistance(neighbor_pos, goal),
                    current
                );
                Node* neighbor_ptr = neighbor_node.get();
                node_map[neighbor_pos] = neighbor_ptr;
                open_set.push(neighbor_ptr);
                all_nodes.push_back(std::move(neighbor_node));
                result.explored.push_back(neighbor_pos);
            }
        }
    }
    
    // Reconstruct path if goal was reached
    if (goal_node) {
        result.path = reconstructPath(goal_node);
    }
    
    return result;
}

std::vector<Vec2i> AStar::getNeighbors(Vec2i pos) const {
    std::vector<Vec2i> neighbors;
    
    // 4-directional movement (up, down, left, right)
    const int dx[] = {0, 0, -1, 1};
    const int dy[] = {-1, 1, 0, 0};
    
    for (int i = 0; i < 4; i++) {
        int nx = pos.x + dx[i];
        int ny = pos.y + dy[i];
        
        if (grid_.isValid(nx, ny) && !grid_.isObstacle(nx, ny)) {
            neighbors.push_back(Vec2i(nx, ny));
        }
    }
    
    return neighbors;
}

std::vector<Vec2i> AStar::reconstructPath(Node* goal) const {
    std::vector<Vec2i> path;
    Node* current = goal;
    
    while (current != nullptr) {
        path.push_back(current->pos);
        current = current->parent;
    }
    
    std::reverse(path.begin(), path.end());
    return path;
}

float AStar::euclideanDistance(Vec2i a, Vec2i b) {
    float dx = static_cast<float>(a.x - b.x);
    float dy = static_cast<float>(a.y - b.y);
    return std::sqrt(dx * dx + dy * dy);
}

float AStar::manhattanDistance(Vec2i a, Vec2i b) {
    return static_cast<float>(std::abs(a.x - b.x) + std::abs(a.y - b.y));
}
